### ISSUE TITLE
Question about bias absorption

### ISSUE THREAD

**Issue by jiangwx** [jakc4103/DFQ#16]
Hi Jakc, 
Thank you for your implementation!
I'm trying to reimplement the data free quantization and I'm confused about bias absorbing.
In the implementation of bias absorption with batchnorm layer, I think `bn1.bias -= c`, and `bn2.bias += conv2.weight*bn2.weight/sqrt(bn2.var)*c`, However, in your code, I saw `bn2.bias += conv2.weight*c`. 
I'm wondering why you did not add `bn2.weight/sqrt(bn2.var)`. And here is my implementation:
```python
def bias_absorption(W1, W2, B1, B2, V2, G1, G2, N=3): # W:weight, B:bias, V:variance, G:gamma
    c = B1-N*G1 # beta-3*gamma
    c = c.clamp(0)

    Wc = torch.zeros(W2.shape[0])
    if(W1.shape[0]==W2.shape[1]): # pixelwise convolution
        for i in range(W2.shape[0]):
            Wc[i] = torch.sum(W2[i].squeeze()*c)*G2[i]/torch.sqrt(V2[i])
    else:                         # depthwise convolution
        for i in range(W2.shape[0]):
            Wc[i] = torch.sum(W2[i])*c[i]*G2[i]/torch.sqrt(V2[i])
    for i in range(len(c)):
        B1[i] = B1[i] - c[i]
    for i in range(len(Wc)):
        B2[i] = B2[i] + Wc[i]
    return c
```

In the paper, the author gives the derivation of `c` by the following equations,
![image](https://user-images.githubusercontent.com/37491779/78968579-8ef41500-7b37-11ea-9337-84666b3bff8f.png)
and thus
![image](https://user-images.githubusercontent.com/37491779/78968654-bfd44a00-7b37-11ea-926e-fb82040ee5fe.png)

However, the equation above neglect the batchnorm, if we take batchnorm into account, we will have:
![image](https://user-images.githubusercontent.com/37491779/78968677-ca8edf00-7b37-11ea-8e6a-7800e8f631f1.png)

where `$\mathbf{W},\gamma, \beta, \mu, \operatorname{Var}$` is the origin paramter of convolution and batchnorm.

For ease of description, the equation above is reduced to 
![image](https://user-images.githubusercontent.com/37491779/78968702-d5e20a80-7b37-11ea-818a-434d9251915d.png)

Thus, we have
![image](https://user-images.githubusercontent.com/37491779/78968718-de3a4580-7b37-11ea-8412-ad65b81bef72.png)

Thank you very much!

**Comment by jakc4103** [comment]
I fuse BN before bias absorption, so by the time doing absorption, BN is just an identity layer.
But you are right, the formula from paper just ignore the BN bias

**Comment by jiangwx** [comment]
Thank you!

**Comment by jiangwx** [comment]
@jakc4103
 Thank you!
